be/src/exprs/ngram.cpp

```diff
@@ -0,0 +1,215 @@
+// Copyright 2021-present StarRocks, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// This file is based on code available under the Apache license here:
+//  https://github.com/ClickHouse/ClickHouse/blob/master/src/Functions/FunctionsStringSimilarity.cpp
+
+#include "column/column_hash.h"
+#include "exprs/string_functions.h"
+namespace starrocks {
+static constexpr size_t max_string_size = 1 << 15;
+
+template <size_t N, bool case_insensitive, bool use_utf_8, class Gram>
+class NgramFunctionImpl {
+public:
+    // we restrict needle's size smaller than 2^16, so even if every gram in needle is the same as each other
+    // we still only need one uint16 to store its frequency
+    using NgramHash = uint16;
+    // uint16[2^16] can almost fit into L2
+    static constexpr size_t map_size = 1 << 16;
+
+    static NgramHash getAsciiHash(const Gram* ch) { return crc_hash_32(ch, N, CRC_HASH_SEED1) & (0xffffu); }
+
+    // for every gram of needle, we calculate its' hash value and store its' frequency in map, and return the number of gram in needle
+    size_t static calculateMapWithNeedle(NgramHash* map, const Slice& needle) {
+        size_t needle_length = needle.get_size();
+        NgramHash cur_hash;
+        size_t i;
+        Slice cur_needle(needle.get_data(), needle_length);
+        const char* ptr;
+        std::string buf;
+        if constexpr (case_insensitive) {
+            tolower(needle, buf);
+            cur_needle = Slice(buf.c_str(), buf.size());
+        }
+        ptr = cur_needle.get_data();
+
+        for (i = 0; i + N <= needle_length; i++) {
+            cur_hash = getAsciiHash(ptr + i);
+            map[cur_hash]++;
+        }
+
+        return i;
+    }
+
+    // traverse haystack‘s every gram, find whether this gram is in needle or not using gram's hash
+    // 16bit hash value may cause hash collision, but because we just calculate the similarity of two string
+    // so don't need to be so accurate.
+    template <bool need_recovery_map>
```

Comment:
what's the meaning of recovery?

Reply:
recovery means: for one row of haystack,we will modify hash table, so for next row of haystack ,we have to recocery the hash table to the original state( hash table filled by calculateMapWithNeedle()).

Reply:
we only have one hashmap for needle， and calculate |(S(needle) - S(haystack))| by traversing haystack's ngram and Subtract 1 from the corresponding item in hash map if hit hashmap. so after handle one haystack, we need recovery the hashmap by add 1 from the corresponding item.